### PR TITLE
[SPARK-26021][SQL][followup] only deal with NaN and -0.0 in UnsafeWriter

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

A followup of https://github.com/apache/spark/pull/23043

There are 4 places we need to deal with NaN and -0.0:
1. comparison expressions. `-0.0` and `0.0` should be treated as same. Different NaNs should be treated as same. This should be true for prefix comparator as well, https://github.com/apache/spark/pull/23334 fixes it.
2. Join keys. `-0.0` and `0.0` should be treated as same. Different NaNs should be treated as same.
3. grouping keys. `-0.0` and `0.0` should be assigned to the same group. Different NaNs should be assigned to the same group.
4. window partition keys. `-0.0` and `0.0` should be treated as same. Different NaNs should be treated as same.

The case 1 is OK. Our comparison already handles NaN and -0.0, and for struct/array/map, we will recursively compare the fields/elements.

Case 2, 3 and 4 are problematic, as they compare `UnsafeRow` binary directly, and different NaNs have different binary representation, and the same thing happens for -0.0 and 0.0.

To fix it, a simple solution is: normalize float/double when building unsafe data (`UnsafeRow`, `UnsafeArrayData`, `UnsafeMapData`). Then we don't need to worry about it anymore.

Following this direction, this PR moves the handling of NaN and -0.0 from `Platform` to `UnsafeWriter`, so that places like `UnsafeRow.setFloat` will not handle them, which reduces the perf overhead. It's also easier to add comments explaining why we do it in `UnsafeWriter`.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @adoron @kiszk @viirya @gatorsmile

**Comment by adoron** [comment]
@cloud-fan what about UnsafeRow::setDouble/Float? It doesn't go through the same flow. Is it not used?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java [hunk 1]
```diff
@@ -174,11 +174,6 @@ public static float getFloat(Object object, long offset) {
   }
 
   public static void putFloat(Object object, long offset, float value) {
-    if (Float.isNaN(value)) {
-      value = Float.NaN;
-    } else if (value == -0.0f) {
-      value = 0.0f;
-    }
```
Comment (dongjoon-hyun) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java]: These change are expected to cause the following test case failure in `PlatformUtilSuite`, but it seems to be missed. Could you fix the test case or remove together, @cloud-fan ?
- https://github.com/apache/spark/blob/master/common/unsafe/src/test/java/org/apache/spark/unsafe/PlatformUtilSuite.java#L162-L163

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java [hunk 2]
```diff
@@ -198,11 +198,45 @@ protected final void writeLong(long offset, long value) {
     Platform.putLong(getBuffer(), offset, value);
   }
 
+  // We need to take care of NaN and -0.0 in several places:
+  //   1. When compare values, different NaNs should be treated as same, `-0.0` and `0.0` should be
+  //      treated as same.
+  //   2. In range partitioner, different NaNs should belong to the same partition, -0.0 and 0.0
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java]: Do we already have related test for case 2?

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java]: It turns out this is not a problem. The doc of `RangePartitioning` is misleading. I'm updating the doc at https://github.com/apache/spark/pull/23249

**Comment by cloud-fan** [comment]
retest this please

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeWriter.java]: As this is not a problem, we should update the PR description too.

**Comment by cloud-fan** [comment]
Yes it is. `UnsafeProjection` always normalize NaN and -0.0, and Spark uses `UnsafeProjection` to produce output. So users can't distinguish them.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by viirya — APPROVED** [review]

**Comment by kiszk** [comment]
The change looks fine.
Do we already have tests for cases 2 and 4?  We know test for case 3 is [here](https://github.com/apache/spark/pull/23043).

**Comment by cloud-fan** [comment]
I checked the original PR that handles NaN: https://github.com/apache/spark/commit/c032b0bf92130dc4facb003f0deaeb1228aefded

It didn't add end-to-end tests, so I added 2 new tests.

**Comment by kiszk** [comment]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]
It seems that all comments are addressed.
Merged to master.

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . Please make another PR for `branch-2.4`. There is a conflict on `branch-2.4`.

**Pull request closed without merging**